### ISSUE TITLE
Warn when using `format!` or `Vec::join` in `fmt::Display` and `fmt::Debug` implementations

### ISSUE THREAD

**Issue by demosdemon** [rust-lang/rust-clippy#15224]
### What it does

Within an implementation of `core::fmt::<Various>::fmt`, it is usually wasteful and unintentional to use `format!` or `Vec::from_iterator` + `Vec::join` and other similar allocating methods.

This lint would warn on such cases to educate the author on ways to prevent allocating in the `fmt` method.

Similar warnings would be useful for manual implementations of `serde::Serialize`, but is out of scope for this warning.

### Advantage

- Removes wasteful alloc/free during `Debug/Display::fmt` impacting performance.

### Drawbacks

There are situations where an author may need to allocate an intermediate string before writing to the formatter such as 
- when handling width and padding inputs. In such cases, this lint would be a false positive. 
- or if the underlying writer for the formatter is not well suited to handle short-writes, then the suggested changes may worsen performance.

Additionally, the recommended code is admittedly harder to read. 

### Example

```rust
struct DisplayMySlice<'a>(&'a [ComplexObject]);

impl std::fmt::Display for DisplayMySlice<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "[{}]",
            self.0
                .iter()
                .map(|x| format!("{x}"))
                .collect::<Vec<_>>()
                .join(", ")
        )
    }
}
```

Could be written as:

```rust
impl std::fmt::Display for DisplayMySlice<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut iter = self.0.iter();
        if let Some(value) = iter.next() {
            write!(f, "[{value}")?;
            for value in iter {
                write!(f, ", {value}")?;
            }
            write!(f, "]")
        } else {
            write!(f, "[]")
        }
    }
}
```

When [`iter::Intersperse`](https://github.com/rust-lang/rust/issues/79524) is stabilized, the recommendation would be to cast the values to `&dyn Display` and use that instead:

```rust
#![feature(iter_intersperse)]
impl std::fmt::Display for DisplayMySlice<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for v in self
            .0
            .iter()
            .map(|v| v as &dyn std::fmt::Display)
            .intersperse(&", ")
        {
            write!(f, "{v}")?;
        }
        Ok(())
    }
}
```
